### PR TITLE
Reuse existing metrics graph

### PR DESCRIPTION
**Pull request by zeapoz**
Now creates `Graph` only once and updates it progressively.

### CODE REVIEW

File: src/tools/crawler/metrics.rs [hunk 1]
```diff
@@ -4,13 +4,54 @@ use std::{
...
+        let (drained, connections): (Vec<_>, Vec<_>) = crawler.known_network.connections()
+            .into_iter()
+            .partition(|&conn| {
+                conn.last_seen.elapsed().as_secs() > LAST_SEEN_CUTOFF
+            });
+
+        // Add new connections.
+        for conn in connections {
+            let edge = Edge::new(conn.a, conn.b);
+            if !(self.graph.contains(&edge)) {
```
Comment (ljedrz) [inline src/tools/crawler/metrics.rs:38]: we shouldn't do a lookup here, as the `Edges` are identical (so replacing an existing one with a new one does nothing); it's faster to just do `self.graph.insert(edge);` and not care

File: src/tools/crawler/metrics.rs [hunk 2]
```diff
@@ -28,7 +69,7 @@ pub struct NetworkSummary {
 
 impl NetworkSummary {
     /// Constructs a new NetworkSummary from given nodes.
-    pub fn new(crawler: &Crawler) -> NetworkSummary {
+    pub fn new(crawler: &Crawler, graph: &mut Graph<SocketAddr>) -> NetworkSummary {
```
[the remark this answers is not in the document]
Comment (ljedrz) [reply src/tools/crawler/metrics.rs:72]: it actually _is_ needed :detective:

**Review by ljedrz — COMMENTED** [review]
LGTM in general, left some comments

File: src/tools/crawler/metrics.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (niklaslong) [reply src/tools/crawler/metrics.rs:62]: It currently is but as discussed, this is something that should be changed in `spectre`.

File: src/tools/crawler/network.rs [hunk 3]
```diff
@@ -125,9 +125,7 @@ impl KnownNetwork {
 
         let mut nodes = self.nodes.write();
         for addr in prospect_nodes {
-            if !nodes.contains_key(&addr) {
-                nodes.insert(addr, KnownNode::default());
-            }
+            nodes.entry(addr).or_insert_with(KnownNode::default);
```
Comment (ljedrz) [inline src/tools/crawler/network.rs:128]:
```suggestion
            nodes.entry(addr).or_default();
```
wouldn't just this one work here :detective:?

File: src/tools/crawler/metrics.rs [hunk 4]
```diff
@@ -4,13 +4,45 @@ use std::{
...
+pub struct NetworkMetrics {
+    graph: Graph<SocketAddr>,
+}
+
+impl NetworkMetrics {
+    /// Constructs a new NetworkMetrics with an empty graph.
+    pub fn new() -> NetworkMetrics {
+        NetworkMetrics::default()
+    }
+
```
Comment (ljedrz) [inline src/tools/crawler/metrics.rs:28]:
```suggestion
```

File: src/tools/crawler/main.rs [hunk 5]
```diff
@@ -58,6 +58,8 @@ async fn main() {
     // Create the crawler with the given listener address.
     let crawler = Crawler::new().await;
 
+    let mut network_metrics = NetworkMetrics::new();
```
Comment (ljedrz) [inline src/tools/crawler/main.rs:61]:
```suggestion
    let mut network_metrics = NetworkMetrics::default();
```

**Review by ljedrz — APPROVED** [review]
LGTM :ok_hand:

**Pull request merged**